### PR TITLE
[service] Add a reusable pool for backend services.

### PR DESCRIPTION
**Pull request by ChrisCummins**
**tldr;** Reuse backend services rather than throwing them away, reducing the cost of `gym.make(..)` from **109**ms-**662**ms to **0.4**ms-**1.4**ms (**84x** to **>1000x** speedups).

## Background

CompilerGym environments use a client/service architecture, where the client is a `gym.Env` object and the service is a subproces that implements the CompilationSession gRPC interface. At the moment, a new service is created every time an environment is created using `gym.make(...)`, and destroyed when the environment is closed. Starting a service is an expensive process (approx 112ms for LLVM and 600ms for loop_tool and GCC), and services can be re-used and shared between clients. There are even improvements to doing so as it means that caches can be shared between clients.

## Overview

This PR adds a `ServiceConnectionPool` class that implements a thread-safe pool for compiler service connections. This enables compiler service connections to be reused, avoiding the expensive initialization of a new service.

There is a global instance of this class, available via the static `ServiceConnectionPool.get()` method. To use the pool, acquire a reference to the global instance, and call the `ServiceConnectionPool.acquire()` method to construct and return service connections:

```py
>>> pool = ServiceConnectionPool.get()
>>> with pool.acquire(Path("/path/to/service"), ConnectionOpts()) as service:
...    # Do something with the service.
```

When a service is closed (by calling `service.close()`), it is automatically released back to the pool so that a future request for the same type of service will reuse the connection.

## Performance

By amortizing the cost of service initialization, the cost of `gym.make(...)` is just that of initializing the relevant `CompilerEnv` class. For "dummy" C++ and Python environments, the speedup is **310x** and **>1000x**, respectively. For the LLVM environment, which has an expensive class initialization, the speedup is **84.4x**.

Detailed microbenchmark results, compared to development branch:

```
------------------------------------------------------- benchmark 'test_make_env[dummy-cc]': 2 tests ------------------------------------------------------
Name (time in ms)                         Min              Median                 Max                Mean            StdDev                   OPS          
-----------------------------------------------------------------------------------------------------------------------------------------------------------
test_make_env[dummy-cc] (dev)        107.1154 (336.40)   108.6546 (320.97)   115.6805 (81.63)    108.8695 (310.03)   1.2997 (11.89)        9.1853 (0.00)   
test_make_env[dummy-cc] (pr-583)       0.3184 (1.0)        0.3385 (1.0)        1.4172 (1.0)        0.3512 (1.0)      0.1093 (1.0)      2,847.7579 (1.0)    
-----------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------- benchmark 'test_make_env[dummy-py]': 2 tests ------------------------------------------------------
Name (time in ms)                         Min              Median                 Max                Mean            StdDev                   OPS          
-----------------------------------------------------------------------------------------------------------------------------------------------------------
test_make_env[dummy-py] (dev)        609.9368 (>1000.0)  661.8493 (>1000.0)  664.0083 (>1000.0)  661.3323 (>1000.0)  5.2601 (125.41)       1.5121 (0.00)   
test_make_env[dummy-py] (pr-583)       0.3381 (1.0)        0.3533 (1.0)        0.5634 (1.0)        0.3753 (1.0)      0.0419 (1.0)      2,664.3997 (1.0)    
-----------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------ benchmark 'test_make_env[llvm]': 2 tests -----------------------------------------------------
Name (time in ms)                     Min              Median                 Max                Mean            StdDev                 OPS          
-----------------------------------------------------------------------------------------------------------------------------------------------------
test_make_env[llvm] (dev)        108.6013 (88.99)    112.1703 (88.07)    153.3195 (90.20)    112.7894 (84.42)    4.4759 (33.44)      8.8661 (0.01)   
test_make_env[llvm] (pr-583)       1.2204 (1.0)        1.2736 (1.0)        1.6997 (1.0)        1.3361 (1.0)      0.1339 (1.0)      748.4634 (1.0)    
-----------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------ benchmark 'test_make_env[loop_tool]': 2 tests -------------------------------------------------------
Name (time in ms)                          Min              Median                 Max                Mean            StdDev                   OPS          
------------------------------------------------------------------------------------------------------------------------------------------------------------
test_make_env[loop_tool] (dev)        660.0815 (>1000.0)  661.9509 (>1000.0)  664.2796 (>1000.0)  661.9331 (>1000.0)  0.7219 (19.71)        1.5107 (0.00)   
test_make_env[loop_tool] (pr-583)       0.3407 (1.0)        0.3553 (1.0)        0.5992 (1.0)        0.3649 (1.0)      0.0366 (1.0)      2,740.4397 (1.0)    
------------------------------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by ChrisCummins** [comment]
Rebased on development

**Review by mostafaelhoushi — APPROVED** [review]
LGTM (skimmed through the code, didn't understand it all but got some idea )
